ko/build/gobuild.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/google/go-containerregistry/v1"
+	"github.com/google/go-containerregistry/v1/mutate"
+	"github.com/google/go-containerregistry/v1/tarball"
+	"github.com/google/go-containerregistry/v1/v1util"
+)
+
+var (
+	// getwd is a variable for testing.
+	getwd = os.Getwd
+)
+
+const (
+	appPath = "/app"
+)
+
+type Options struct {
+	// TODO(mattmoor): Base Image?
+	// TODO(mattmoor): Architectures?
+	Base v1.Image
+}
+
+type gobuild struct {
+	importpath string
+	opt        Options
+	build      func(string) (string, error)
+}
+
+func computeImportpath() (string, error) {
+	wd, err := getwd()
+	if err != nil {
+		return "", err
+	}
+	// Go code lives under $GOPATH/src/...
+	src := path.Join(os.Getenv("GOPATH"), "src")
+	if !strings.HasPrefix(wd, src) {
+		return "", fmt.Errorf("working directory %q must be on GOPATH %q", wd, src)
+	}
+	return strings.Trim(strings.TrimPrefix(wd, src), "/"), nil
+}
+
+// NewGo returns a build.Interface implementation that:
+//  1. builds go binaries named by importpath,
+//  2. containerizes the binary on a suitable base,
```

Comment:
"suitable" -> "specified"

The user might select a totally bonkersgonuts base image, which wouldn't be suitable at all.